### PR TITLE
[llvm-readelf] Add --extra-sym-info

### PR DESCRIPTION
**Pull request by MaskRay**
GNU readelf introduced --extra-sym-info/-X to display the section name
for --syms (https://sourceware.org/PR30684). Port the feature, which is
currently llvm-readelf only.

For STO_AARCH64_VARIANT_PCS/STO_RISCV_VARIANT_PCS, the Ndx and Name
columns may not be aligned.

### CODE REVIEW

File: llvm/test/tools/llvm-readobj/ELF/symbols.test [hunk 1]
```diff
@@ -52,19 +54,44 @@
...
+#      SYMBOLS-GNU32:Symbol table '.symtab' contains 4 entries:
 # SYMBOLS-GNU32-NEXT:   Num:    Value  Size Type    Bind   Vis       Ndx Name
 # SYMBOLS-GNU32-NEXT:     0: 00000000     0 NOTYPE  LOCAL  DEFAULT   UND 
-# SYMBOLS-GNU32-NEXT:     1: 00000001     0 NOTYPE  LOCAL  DEFAULT   UND foo
+# SYMBOLS-GNU32-NEXT:     1: 00000001     0 NOTYPE  LOCAL  DEFAULT     1 foo
 # SYMBOLS-GNU32-NEXT:     2: 00000002     0 NOTYPE  LOCAL  DEFAULT   UND bar
+# SYMBOLS-GNU32-NEXT:     3: 00000003     0 NOTYPE  GLOBAL DEFAULT     2 data
+
+#     SYMBOLS-GNU32X:Symbol table '.symtab' contains 4 entries:
+#SYMBOLS-GNU32X-NEXT:   Num:    Value  Size Type    Bind   Vis+Other Ndx(SecName) Name [+ Version Info]
```
Comment (jh7370) [inline llvm/test/tools/llvm-readobj/ELF/symbols.test:90]: Nit: is there a reason to omit the space entirely between the `#` and the check pattern name? Same goes elsewhere in the file.

File: llvm/test/tools/llvm-readobj/ELF/symbols.test [hunk 2]
```diff
@@ -13,9 +13,11 @@
 # RUN: yaml2obj %s -DBITS=64 -DTYPE=ET_REL -o %t64
 # RUN: llvm-readobj --symbols %t64 | FileCheck %s --match-full-lines --strict-whitespace --check-prefix=SYMBOLS-LLVM
 # RUN: llvm-readelf --symbols %t64 | FileCheck %s --match-full-lines --strict-whitespace --check-prefix=SYMBOLS-GNU64
+# RUN: llvm-readelf -s --extra-sym-info %t64 | FileCheck %s --match-full-lines --strict-whitespace --check-prefix=SYMBOLS-GNU64X
```
Comment (jh7370) [inline llvm/test/tools/llvm-readobj/ELF/symbols.test:16]: Is it worth adding a check showing that --extra-sym-info has no impact on the output for LLVM output format? That way it's a clear indication where any testing belongs, should the behaviour change in the future.

File: llvm/test/tools/llvm-readobj/ELF/aarch64-symbols-stother.test [hunk 3]
```diff
@@ -29,6 +30,15 @@
...
 # GNU-NEXT: 4: 0000000000000000 0 NOTYPE LOCAL PROTECTED                    UND foo4
 
+#      GNUX:Symbol table '.symtab' contains 5 entries:
+# GNUX-NEXT:   Num:    Value          Size Type    Bind   Vis+Other              Ndx(SecName) Name [+ Version Info]
+# GNUX-NEXT:     0: 0000000000000000     0 NOTYPE  LOCAL  DEFAULT                UND 
+# GNUX-NEXT:     1: 0000000000000000     0 NOTYPE  LOCAL  DEFAULT [VARIANT_PCS]  UND foo1
+# GNUX-NEXT:     2: 0000000000000000     0 NOTYPE  LOCAL  DEFAULT [VARIANT_PCS | 40] UND foo2
+# GNUX-NEXT:     3: 0000000000000000     0 NOTYPE  LOCAL  PROTECTED [VARIANT_PCS] UND foo3
+# GNUX-NEXT:     4: 0000000000000000     0 NOTYPE  LOCAL  PROTECTED              UND foo4
+
```
Comment (jh7370) [inline llvm/test/tools/llvm-readobj/ELF/aarch64-symbols-stother.test:40]: Unnecessary double blank line?

File: llvm/tools/llvm-readobj/ELFDumper.cpp [hunk 4]
```diff
@@ -4058,17 +4075,31 @@ GNUELFDumper<ELFT>::getSymbolSectionNdx(const Elf_Sym &Symbol,
...
-    // A normal section with an index
-    return to_string(format_decimal(SectionIndex, 3));
+    break;
   }
+
+  std::string Extra;
+  if (ExtraSymInfo) {
+    auto Sec = this->Obj.getSection(SectionIndex);
+    if (!Sec) {
+      this->reportUniqueWarning(Sec.takeError());
```
Comment (jh7370) [inline llvm/tools/llvm-readobj/ELFDumper.cpp:4085]: Looks like this and the below warning aren't actually tested?

**Review by jh7370 — COMMENTED** [review]
No major comments, all looks generally reasonable to me.

I noticed this option extends the column headers, but we're already printing the extra information without the option. Is that the same as GNU, or are we currently printing more than them?

File: llvm/test/tools/llvm-readobj/ELF/symbols.test (hunk 1 above)
Comment (MaskRay) [reply llvm/test/tools/llvm-readobj/ELF/symbols.test:90]: This is to align with `SYMBOLS-GNU32` above. Happy to indent by 1 if you feel strongly.

File: llvm/tools/llvm-readobj/ELFDumper.cpp (hunk 4 above)
Comment (MaskRay) [reply llvm/tools/llvm-readobj/ELFDumper.cpp:4085]: Added more tests to symbols.test and section-symbols.test

File: llvm/test/tools/llvm-readobj/ELF/symbols.test (hunk 2 above)
Comment (MaskRay) [reply llvm/test/tools/llvm-readobj/ELF/symbols.test:16]: Added

File: llvm/test/tools/llvm-readobj/ELF/aarch64-symbols-stother.test (hunk 3 above)
Comment (MaskRay) [reply llvm/test/tools/llvm-readobj/ELF/aarch64-symbols-stother.test:40]: Removed

**Review by jh7370 — APPROVED** [review]
LGTM! Maybe a release note too?

**Pull request merged**